### DISCUSSION TITLE
[Question]: Is there lag between MQTT & HTTP

### DISCUSSION THREAD

**Discussion by hipitihop** [blakeblackshear/frigate#9957] [Ask A Question] [answered]
### What is your question:

`Background:`
I'm on the latest Frigate 0.13.2-6476F8A docker. I'm experimenting with doing ALPR in multiple zones at the drive gates and also in an internal drive area outside my garage. One to open gates, the other to open garage. 
I'm using the "Plate Recognizer" snapshot SDK, also running in another docker container. 

The current strategy follows:
- A Node Red flow listens for frigate/event MQTT filtered for properties  after/camera and after/label = "car"
- So far I have found the event type "update" most reliable as "new" doesn't have anything in current_zones or entered_zones and I don't want to wait for "end"
- The flow waits until it sees an "update" event with the after/current_zones  includes the designated ALPR zone
- Using the after/frame_time to construct a frigate URL like http://192.168.192.14:15000/api/front/recordings/1708385933.243289/snapshot.png and fetching it.
- on success, send this to plate recognizer and so on.

`Problem:`
What I have found, is quite often the HTTP GET of the snapshot UR, fails with a status 404, however moments later when debugging the failures, if I paste the same url into a browser, sure enough it is there and returns the snapshot of the car in the zone.
So there seems to be some lag between the MQTT event and when the corresponding frame is available in recordings via this HTTP API.
I can obviously add logic to keep retrying  after a 404 until such time it becomes available, but that slows things down, update events keep coming etc etc. 

I wonder, am I using the wrong API to get the associated frame or does this show some underlying bug.

**Comment by NickM-27** [comment]
You are pulling a snapshot from recordings, recording segments are not always ready immediately since they have a segment length of 10 seconds 

I'd recommend just using the object snapshot API or using latest.jpg

**Comment by hipitihop** [comment]
@NickM-27 thanks. I'm pretty sure I've tried snapshot and object latest before, but I will re check the docs and retry. I think what I'm struggling with is correlating the "update" event frame. There seems to be only one event `id` from "new" , "update" and "end". Snapshot and latest API don't seem to provide a way to specify the frame time either. So from all the frames that cause an MQTT event, how can I get the specific snapshot or latest ? I'm sure I'm missing an epiphany here and or the docs. Waiting 10 seconds is a poor user experience, when you're the one sitting in the car waiting for the garage to open. I'd love to get closer to real time, such that I can detect and match the plate within first 5 frames of entering the zone.

**Reply by blakeblackshear** [reply, marked as answer]
If you want real time, I would suggest using the [latest.jpg endpoint](https://docs.frigate.video/integrations/api#get-apicamera_namelatestjpgh300). The endpoint you are using now is the slowest option available.

**Reply by hipitihop** [reply]
Thanks, I'll try and test over a couple of days and report back. How often is latest.jpg updated? I'm concerned that by the time I do the fetch (get) , that more events have updated it, and the car is no longer there. I guess on fast local network that is less likely though

Appreciate the responses and feedback

**Reply by NickM-27** [reply]
latest.jpg is the latest frame so it is updated as often as frames are received (by default detect fps is 5)

**Comment by hipitihop** [comment]
On further reflection, I'm wondering if it would not be better to include the raw frame bytes (the snapshot) as a field on the MQTT event payload ?

I'm thinking, the code that generates the MQTT event, which populates the `before` and `after`, presumably already has access to the current frame triggering the event. Rather than expecting the consumer of the MQTT event to correlate the event and subsequent retrieval of the associated snapshot using another asynchronous API call, the MQTT generating handler could simply attache it.

I understand people are concerned about payload size, etc.  It is actually quite surprising how robust this can be over MQTT. In another project which provides a voice based assistant with STT and TTS, then raw audio bytes are streamed the same way. On a local network, it's a non issue. Alternatively some may argue this should be a config option to minimize the bandwidth MQTT events could consume.

I'm happy to contribute a PR if someone can point me at where the current frame bytes are available while generating the MQTT event.

**Reply by NickM-27** [reply]
I don't think that makes sense. That's a lot of data many times per second especially as it scales across larger setups. I have an automation that runs ALPR and another that runs face processing and I've never had an issue with getting the frame to successfully run these things.

**Reply by hipitihop** [reply]
For the `frigate/events` MQTT topic, from examples I have seen, it does not look like an MQTT message is generated for each frame. I'm seeing something like one event each 5 seconds.  Nonetheless, I'm experimenting with using the latest.jpg API, and so far this indeed seems to be working well. I appreciate everyone's responses and help. Loving Frigate, many thanks. Closing
